supertokens_python/recipe/session/jwks.py

```diff
@@ -1,104 +1,127 @@
+    "cache_max_age": JWKCacheMaxAgeInMs,
+    "request_timeout": 5000,  # 5s
+}
 
-    def get_latest_keys(self) -> List[PyJWK]:
-        if self.jwk_set is None or not self.is_fresh():
-            self.reload()
 
-        if self.jwk_set is None:
-            raise JWKSRequestError("Failed to fetch the latest keys")
+class CachedKeys:
+    def __init__(self, keys: List[PyJWK]):
+        self.keys = keys
+        self.last_refresh_time = get_timestamp_ms()
 
-        all_keys: List[PyJWK] = self.jwk_set.keys  # type: ignore
+    def is_fresh(self):
+        return get_timestamp_ms() - self.last_refresh_time < JWKSConfig["cache_max_age"]
 
-        return all_keys
 
-    def get_matching_key_from_jwt(self, token: str) -> PyJWK:
-        header = decode_token(token, options={"verify_signature": False})["header"]
-        kid: str = header["kid"]  # type: ignore
+cached_keys: Optional[CachedKeys] = None
+mutex = RWMutex()
 
-        if self.jwk_set is None or not self.is_fresh():
-            self.reload()
+# only for testing purposes
+def reset_jwks_cache():
+    with RWLockContext(mutex, read=False):
+        global cached_keys
+        cached_keys = None
 
-        assert self.jwk_set is not None
 
-        try:
-            return self.jwk_set[kid]  # type: ignore
-        except KeyError:
-            if not self.is_cooling_down():
-                # One more attempt to fetch the latest keys
-                # and then try to find the key again.
-                self.reload()
-                try:
-                    return self.jwk_set[kid]  # type: ignore
-                except KeyError:
-                    pass
-        except Exception:
-            raise JWKSKeyNotFoundError("No key found for the given kid")
+def get_cached_keys() -> Optional[CachedKeys]:
+    with RWLockContext(mutex, read=True):
+        if cached_keys is not None:
+            # This means that we have valid JWKs for the given core path
+            # We check if we need to refresh before returning
 
-        raise JWKSKeyNotFoundError("No key found for the given kid")
+            # This means that the value in cache is not expired, in this case we return the cached value
+            # Note that this also means that the SDK will not try to query any other core (if there are multiple)
+            # if it has a valid cache entry from one of the core URLs. It will only attempt to fetch
+            # from the cores again after the entry in the cache is expired
+            if cached_keys.is_fresh():
+                if environ.get("SUPERTOKENS_ENV") == "testing":
+                    log_debug_message("Returning JWKS from cache")
+                return cached_keys
 
+        return None
 
-class JWKSKeyNotFoundError(Exception):
-    pass
 
+def get_latest_keys(kid: Optional[str] = None) -> List[PyJWK]:
+    global cached_keys
+
+    if environ.get("SUPERTOKENS_ENV") == "testing":
+        log_debug_message("Called find_jwk_client")
+
```

Comment:
lock the whole thing below with a read lock

Reply:
Done.